model/util.go

```diff
@@ -19,112 +19,243 @@ import (
+	if e.err == nil {
+		panic("unmarshalError fail")
+	}
 
-// FALSE used by bool fields that needs a boolean pointer
-var FALSE = false
+	switch wrapErr := e.err.(type) {
+	case *json.SyntaxError:
+		message = fmt.Sprintf("%s has a syntax error \"%s\"", e.parameterName, wrapErr.Error())
+	case *json.UnmarshalTypeError:
+		var primitiveTypeName string
+		var objectTypeName string
 
-func getBytesFromFile(s string) (b []byte, err error) {
-	// #nosec
-	if resp, err := http.Get(s); err == nil {
-		defer resp.Body.Close()
-		buf := new(bytes.Buffer)
-		if _, err = buf.ReadFrom(resp.Body); err != nil {
-			return nil, err
+		if e.primitiveType != 0 {
+			primitiveTypeName = e.primitiveType.String()
 		}
-		return buf.Bytes(), nil
-	}
-	s = strings.TrimPrefix(s, prefix)
-
-	if !filepath.IsAbs(s) {
-		// The import file is an non-absolute path, we join it with include path
-		// TODO: if the file didn't find in any include path, we should report an error
-		for _, p := range IncludePaths() {
-			sn := filepath.Join(p, s)
-			if _, err := os.Stat(sn); err == nil {
-				s = sn
-				break
+		if e.objectType != 0 {
+			switch e.objectType {
+			case reflect.Struct:
+				objectTypeName = "object"
+			case reflect.Map:
+				objectTypeName = "object"
+			case reflect.Slice:
+				objectTypeName = "array"
+			default:
+				objectTypeName = e.objectType.String()
 			}
 		}
+
+		if wrapErr.Struct == "" && wrapErr.Field == "" {
+			message = fmt.Sprintf("%s must be an %s or %s", e.parameterName, primitiveTypeName, objectTypeName)
+		} else if wrapErr.Struct != "" && wrapErr.Field != "" {
+
+			switch wrapErr.Type {
+			case reflect.TypeOf(InvokeKindSync):
+				primitiveTypeName = fmt.Sprintf("%s,%s", InvokeKindSync, InvokeKindAsync)
+			case reflect.TypeOf(ForEachModeTypeSequential):
+				primitiveTypeName = fmt.Sprintf("%s,%s", ForEachModeTypeSequential, ForEachModeTypeParallel)
+			default:
+				primitiveTypeName = wrapErr.Type.Name()
+			}
+
+			message = fmt.Sprintf("%s.%s must be an %s", e.parameterName, wrapErr.Field, primitiveTypeName)
+		} else {
+			message = wrapErr.Error()
+		}
+
+	default:
+		message = wrapErr.Error()
 	}
 
-	if b, err = os.ReadFile(filepath.Clean(s)); err != nil {
+	return message
+}
+
+func getBytesFromFile(uri string) (b []byte, err error) {
+	refUrl, err := url.Parse(uri)
+	if err != nil {
 		return nil, err
 	}
 
+	if refUrl.Scheme == "file" {
+		path := filepath.Join(refUrl.Host, refUrl.Path)
+		if !filepath.IsAbs(path) {
+			// The import file is an non-absolute path, we join it with include path
+			// TODO: if the file didn't find in any include path, we should report an error
+			for _, p := range IncludePaths() {
+				sn := filepath.Join(p, path)
+				if _, err := os.Stat(sn); err == nil {
+					path = sn
+					break
+				}
+			}
+		}
+
+		b, err = os.ReadFile(filepath.Clean(path))
+		if err != nil {
+			return nil, err
+		}
+
+	} else {
+		// #nosec
+		req, err := http.NewRequest(http.MethodGet, refUrl.String(), nil)
```

Comment:
what is the purpose of it?

Reply:
To write unit tests with httptest. Example in the `fileunit_test.go` `Test_getBytesFromFile`.

Reply:
yes, but why is it needed?

Reply:
Following the specification rule. "Referencing a reusable JSON/Yaml error definition file" an example the [defining errors](https://github.com/serverlessworkflow/specification/blob/0.8.x/specification.md#defining-errors
) 

What schemes can the URI be: `file://`, `https://`, `ftp://`?

Reply:
I see, but I think it is up to the engine to implement it, no?
@ricardozanini wdyt?

Reply:
No, this is our duty since the external file contains the definition for the object. For example:

```json
{
    "functions": "myfunctions.json"
}
```

The structure should be `Workflow().Functions()[0].Name`. Not `Workflow().Functions()` ~ a string

@ribeiromiranda the scheme should be determined by a URI parser. By default, it's ` file://` if the scheme is empty. If remote like HTTP(S), we do fetch.